Boolean_set_operations_2/include/CGAL/Boolean_set_operations_2/Bso_internal_functions.h

```diff
@@ -8,192 +8,522 @@
-inline bool _join(const Pgn1& pgn1, const Pgn2& pgn2,
-                  typename Traits::Polygon_with_holes_2& res, Traits& tr)
-{
-  if (_is_empty(pgn1, tr) || _is_empty(pgn2, tr))
-    return false;
+//@}
 
-  General_polygon_set_2<Traits> gps(tr);
-  gps.insert(pgn1);
+/// \name s_join() and r_join() functions.
+//@{
+
+// Single operands
+// With Traits
+template <typename Pgn1, typename Pgn2, typename Traits>
+inline bool s_join(const Pgn1& pgn1, const Pgn2& pgn2,
+                   typename Traits::Polygon_with_holes_2& res, Traits& traits) {
+  if (_is_empty(pgn1, traits) || _is_empty(pgn2, traits)) return false;
+
+  General_polygon_set_2<Traits> gps(pgn1, traits);
   gps.join(pgn2);
-  if (gps.number_of_polygons_with_holes() == 1)
-  {
-    Oneset_iterator<typename Traits::Polygon_with_holes_2> oi (res);
+  if (gps.number_of_polygons_with_holes() == 1) {
+    Oneset_iterator<typename Traits::Polygon_with_holes_2> oi(res);
     gps.polygons_with_holes(oi);
     return true;
   }
 
-  // the polygon doesnt intersect, the original pgn1, pgn2 contain the union
+  // The polygon doesn't intersect; the original pgn1, pgn2 contain the union
   return false;
 }
 
-template <class Pgn1, class Pgn2, class Pwh>
-inline bool _join(const Pgn1& pgn1, const Pgn2& pgn2, Pwh& res)
-{
-  typename Gps_default_traits<Pgn1>::Traits  tr;
-  return _join(pgn1, pgn2, res, tr);
+// Without Traits
+template <typename Kernel, typename Container,
+          typename Pgn1, typename Pgn2, typename Pwh>
+inline bool s_join(const Pgn1& pgn1, const Pgn2& pgn2, Pwh& pwh) {
+  // Use the first polygon to determine the (default) traits
+  typedef typename Gps_polyline_traits<Pgn1>::Polyline_traits   Polyline_traits;
+  typedef General_polygon_2<Polyline_traits>                    General_pgn;
+  typedef General_polygon_with_holes_2<General_pgn>             General_pwh;
+
+  General_pwh general_pwh;
+  typename Gps_polyline_traits<Pgn1>::Traits traits;
+  const Polyline_traits& ptraits(traits);
+  auto res = s_join(convert_polygon(pgn1, ptraits),
+                    convert_polygon(pgn2, ptraits),
+                    general_pwh, traits);
+  pwh = convert_polygon_back<Kernel, Container>(general_pwh);
+  return res;
+}
+
+// Range
+// With traits
+template <typename InputIterator, typename OutputIterator, typename Traits>
+inline OutputIterator r_join(InputIterator begin, InputIterator end,
+                             OutputIterator oi, Traits& traits,
+                             unsigned int k=5) {
+  if (begin == end) return oi;
+  General_polygon_set_2<Traits> gps(*begin, traits);
+  gps.join(std::next(begin), end, k);
+  return gps.polygons_with_holes(oi);
+}
+
+// Without traits
+template <typename InputIterator, typename OutputIterator>
+inline OutputIterator r_join(InputIterator begin, InputIterator end,
+                             OutputIterator oi, unsigned int k=5) {
+  typedef typename std::iterator_traits<InputIterator>::value_type Pgn;
```

Comment:
Here it needs the same `if (begin == end) return oi;` as in line 284, as otherwise the `*begin` in line 299 fails.

Reply:
I am working on the patch

Reply:
It seems your commit and my patch are in conflict.
Let me know if you need a hand for msvc 2017

Reply:
So maybe just review my fix.   It now passes locally for me.

Reply:
I should have confirmed Sebastien that I take the task.